File: claude/harness/steambench_harness/envs/catcher.py
"""Catcher — a deterministic SteamBench arcade *vision* env (richer than Dodger).

Slide a paddle to **catch** the good items (one colour) while **dodging** the bad
ones (another colour). It is the second pixel-runtime game: where Dodger only
asks "is this column dangerous?", Catcher forces a *two-class* perception — tell
good from bad, then approach one and avoid the other — which is exactly the
harder thing a real vision agent must do. Shared bit-for-bit with the pixel
runtime so the CV agent's pixel run replay-verifies on this engine.

Randomness (item spawn column, then item kind) comes from ``self.rng``, two draws
per spawn in a fixed order, so ``(seed, actions)`` traces replay across Py/TS.
"""

from __future__ import annotations

from typing import Union

from ..protocol import AchievementSpec, ActionSpace, Env, Observation, VerifyMode
from ..registry import register

# Shared with runtime/pixel_game.py CatcherGame — keep in lock-step (parity test).
W, H = 168, 120
PW, PH = 24, 8          # paddle (a bit wide — it must catch, not just dodge)
IW, IH = 12, 12         # item block
PLAYER_Y = H - PH - 3   # 109
PSPEED = 7
FALL = 4
SPAWN_GAP = 22
BAD_NUM, BAD_DEN = 37, 100  # ~37% of items are bad; integer compare keeps it portable
_ACTIONS = ["left", "stay", "right"]


@register
class Catcher(Env):
    env_id = "arcade/catcher"
    appid = 9000010
    name = "Catcher"
    description = (
        "Catch the falling green drops, dodge the red ones. A two-class vision "
        "task: tell good from bad, sweep up the good, flinch from the bad — one "
        "red touch ends the run. The second game a vision agent plays from pixels."
    )
    verify_mode = VerifyMode.REPLAY
    action_space = ActionSpace(_ACTIONS)
    achievements = [
        AchievementSpec("catch_5", "First Drops", "Catch 5 good items.", 0.6),
        AchievementSpec("catch_15", "Bucket Hands", "Catch 15 good items.", 0.32),
        AchievementSpec("catch_30", "Sticky Fingers", "Catch 30 good items.", 0.13),
        AchievementSpec("catch_60", "Vacuum", "Catch 60 good items.", 0.04),
        AchievementSpec("catch_120", "Event Horizon", "Catch 120 good items.", 0.006),
    ]
    _LADDER = [(5, "catch_5"), (15, "catch_15"), (30, "catch_30"),
               (60, "catch_60"), (120, "catch_120")]

    def __init__(self) -> None:
        super().__init__()
        self.px = (W - PW) // 2
        self.items: list[list[int]] = []  # [x, y, kind] kind 0=good 1=bad
        self.caught = 0
        self.alive = True
        self._since_spawn = SPAWN_GAP

    def reset(self, seed: int = 0) -> Observation:
        self._begin(seed)
        self.px = (W - PW) // 2
        self.items = []
        self.caught = 0
        self.alive = True
        self._since_spawn = SPAWN_GAP
        return self._observe(0.0)

    def step(self, action: Union[int, str]) -> Observation:
        if self.done:
            return self._observe(0.0)
        name = self.action_space.name(action)
        self.steps += 1

        if name == "left":
            self.px = max(0, self.px - PSPEED)
        elif name == "right":
            self.px = min(W - PW, self.px + PSPEED)

        # Spawn (two draws: column, then kind), then fall.
        self._since_spawn += FALL
        if self._since_spawn >= SPAWN_GAP:
            self._since_spawn = 0
            x = self.rng.randrange(W - IW + 1)
            kind = 1 if self.rng.randrange(BAD_DEN) < BAD_NUM else 0
            self.items.append([x, -IH, kind])
        for it in self.items:
            it[1] += FALL

        reward = 0.0
        survivors: list[list[int]] = []
        for x, y, kind in self.items:
            overlaps = (x < self.px + PW and x + IW > self.px and
                        y < PLAYER_Y + PH and y + IH > PLAYER_Y)
            if overlaps:
                if kind == 1:           # caught a bad one — run ends
                    self.alive = False
                    self.done = True
                    reward = -1.0
                else:                    # caught a good one
                    self.caught += 1
                    reward += 1.0
                continue                 # consumed either way
            if y >= H:
                continue                 # fell past — missed
            survivors.append([x, y, kind])
        self.items = survivors

        self.score = self.caught  # always reflects goods caught, even on a death tick
        if not self.done:
            for thresh, aid in self._LADDER:
                if self.caught >= thresh:
                    self._unlock(aid)
        return self._observe(reward)

    def legal_actions(self) -> list[str]:
        return list(_ACTIONS)

    def render(self) -> str:
        goods = sum(1 for it in self.items if it[2] == 0)
        bads = len(self.items) - goods
        return (f"caught={self.caught} px={self.px} good={goods} bad={bads} "
                f"{'ALIVE' if self.alive else 'DEAD'}")

    def _observe(self, reward: float) -> Observation:
        return Observation(
            step=self.steps,
            state={
                "paddle_x": self.px,
                "paddle_y": PLAYER_Y,
                "paddle_w": PW,
                "paddle_h": PH,
                "paddle_speed": PSPEED,
                "item_w": IW,
                "item_h": IH,
                "width": W,
                "height": H,
                "fall": FALL,
                "items": [{"x": x, "y": y, "kind": k} for x, y, k in self.items],
                "caught": self.caught,
                "score": self.caught,
                "alive": self.alive,
            },
            text=self.render(),
            legal_actions=self.legal_actions(),
            score=float(self.caught),
            done=self.done,
            reward=float(reward),
            info={"caught": self.caught, "alive": self.alive},
        )

File: claude/web/src/lib/arcade/catcher.ts
/**
 * Catcher — a deterministic SteamBench arcade *vision* env (richer than Dodger).
 *
 * TypeScript mirror of `harness/steambench_harness/envs/catcher.py` (and the
 * headless twin of the pixel runtime's CatcherGame). Catch the good drops, dodge
 * the bad ones — a two-class perception task.
 *
 * Determinism note: each spawn draws TWO values in a fixed order — the column
 * `rng.randrange(W - IW + 1)`, then the kind `rng.randrange(BAD_DEN) < BAD_NUM`.
 * Everything else is integer arithmetic in Python's exact order, so a recorded
 * `(seed, actions)` trace — human, Python agent, or the vision agent reading
 * pixels — replays to the same score in either language.
 */

import { ActionSpace, Env } from "./base";
import type { AchievementSpec, Action, Observation } from "./types";

const W = 168;
const H = 120;
const PW = 24;
const PH = 8;
const IW = 12;
const IH = 12;
const PLAYER_Y = H - PH - 3; // 109
const PSPEED = 7;
const FALL = 4;
const SPAWN_GAP = 22;
const BAD_NUM = 37;
const BAD_DEN = 100;
const ACTIONS = ["left", "stay", "right"] as const;

export class Catcher extends Env {
  static env_id = "arcade/catcher";
  static appid = 9000010;
  static displayName = "Catcher";
  static description =
    "Catch the falling green drops, dodge the red ones. A two-class vision " +
    "task: tell good from bad, sweep up the good, flinch from the bad — one " +
    "red touch ends the run. The second game a vision agent plays from pixels.";
  static verify_mode = "replay" as const;
  static action_space = new ActionSpace([...ACTIONS]);
  static achievements: AchievementSpec[] = [
    { id: "catch_5", name: "First Drops", description: "Catch 5 good items.", rarity_hint: 0.6 },
    { id: "catch_15", name: "Bucket Hands", description: "Catch 15 good items.", rarity_hint: 0.32 },
    { id: "catch_30", name: "Sticky Fingers", description: "Catch 30 good items.", rarity_hint: 0.13 },
    { id: "catch_60", name: "Vacuum", description: "Catch 60 good items.", rarity_hint: 0.04 },
    { id: "catch_120", name: "Event Horizon", description: "Catch 120 good items.", rarity_hint: 0.006 },
  ];
  private static LADDER: [number, string][] = [
    [5, "catch_5"], [15, "catch_15"], [30, "catch_30"], [60, "catch_60"], [120, "catch_120"],
  ];

  px: number;
  items: [number, number, number][]; // [x, y, kind] kind 0=good 1=bad
  caught: number;
  alive: boolean;
  private sinceSpawn: number;

  constructor() {
    super();
    this.px = Math.trunc((W - PW) / 2);
    this.items = [];
    this.caught = 0;
    this.alive = true;
    this.sinceSpawn = SPAWN_GAP;
  }

  reset(seed = 0): Observation {
    this.begin(seed);
    this.px = Math.trunc((W - PW) / 2);
    this.items = [];
    this.caught = 0;
    this.alive = true;
    this.sinceSpawn = SPAWN_GAP;
    return this.observe(0.0);
  }

  step(action: Action): Observation {
    if (this.done) {
      return this.observe(0.0);
    }
    const name = this.actionSpace.name(action);
    this.steps += 1;

    if (name === "left") {
      this.px = Math.max(0, this.px - PSPEED);
    } else if (name === "right") {
      this.px = Math.min(W - PW, this.px + PSPEED);
    }

    this.sinceSpawn += FALL;
    if (this.sinceSpawn >= SPAWN_GAP) {
      this.sinceSpawn = 0;
      const x = this.rng.randrange(W - IW + 1);
      const kind = this.rng.randrange(BAD_DEN) < BAD_NUM ? 1 : 0;
      this.items.push([x, -IH, kind]);
    }
    for (const it of this.items) {
      it[1] += FALL;
    }

    let reward = 0.0;
    const survivors: [number, number, number][] = [];
    for (const [x, y, kind] of this.items) {
      const overlaps = x < this.px + PW && x + IW > this.px && y < PLAYER_Y + PH && y + IH > PLAYER_Y;
      if (overlaps) {
        if (kind === 1) {
          this.alive = false;
          this.done = true;
          reward = -1.0;
        } else {
          this.caught += 1;
          reward += 1.0;
        }
        continue;
      }
      if (y >= H) {
        continue;
      }
      survivors.push([x, y, kind]);
    }
    this.items = survivors;

    this.score = this.caught; // always reflects goods caught, even on a death tick
    if (!this.done) {
      for (const [thresh, aid] of Catcher.LADDER) {
        if (this.caught >= thresh) {
          this.unlock(aid);
        }
      }
    }
    return this.observe(reward);
  }

  legalActions(): string[] {
    return [...ACTIONS];
  }

  render(): string {
    let goods = 0;
    for (const it of this.items) if (it[2] === 0) goods += 1;
    const bads = this.items.length - goods;
    return `caught=${this.caught} px=${this.px} good=${goods} bad=${bads} ${this.alive ? "ALIVE" : "DEAD"}`;
  }

  private observe(reward: number): Observation {
    return {
      step: this.steps,
      state: {
        paddle_x: this.px,
        paddle_y: PLAYER_Y,
        paddle_w: PW,
        paddle_h: PH,
        paddle_speed: PSPEED,
        item_w: IW,
        item_h: IH,
        width: W,
        height: H,
        fall: FALL,
        items: this.items.map(([x, y, kind]) => ({ x, y, kind })),
        caught: this.caught,
        score: this.caught,
        alive: this.alive,
      },
      text: this.render(),
      frame: null,
      legal_actions: this.legalActions(),
      score: this.caught,
      done: this.done,
      reward: reward,
      info: { caught: this.caught, alive: this.alive },
    };
  }
}
